### DISCUSSION TITLE
How to streaming parse output of external command while it's running?

### DISCUSSION THREAD

**Discussion by cshuaimin** [nushell/nushell#5521] [Q&A] [answered]
```shell
nu -c '0..10 | each {|i| ^echo $i}; sleep 5sec' | lines
```

I want to parse lines while the external command is running, but it seems that I will get all the lines when it exits.

I'm trying to parse the output of a long-running `gdbus monitor` command and get event notifications of screen locks, so it's important to get every line in time.

**Comment by fdncred** [comment, marked as answer]
Dejay in discord figured this out. It's not what I'd call working but it does something. `ping 8.8.8.8 | each {|i| print ($i | parse -r "(?P<time>time=)(?P<ms>.*)")}`. Maybe you can take this and build upon it?

**Reply by cshuaimin** [reply]
Thank you, using `each` + `print` on byte stream works!

However because `print` outputs to stdout and is a Nu builtin command, I can't pipe the output of `print` back. I found `^echo` is a better choice here, since it's an extern command and its output can be piped to builtin commands.

Share with you my final code, which resets the  backlight when I unlock the screen (backlight is set to max due to some driver issues).

```shell
let file = '/sys/class/backlight/acpi_video0/brightness'
let backlight = (open $file)

gdbus monitor -y -d org.freedesktop.login1
| each { |line|
  ^echo ($line | parse -r ''LockedHint': <(?P<locked>\w+)>' | get 0.locked)
  | into bool
  | if not $in {
    ^echo $'restoring backlight to ($backlight)'
    sudo nu -c $'"($backlight)" | save ($file)'
  }
}
```

**Comment by GrabbenD** [comment]
Here's another example to accomplish this _(collected from an interesting discussion between @collide_star and @Bahex in NuShell's Discord)_:

`$ nano ./bash_script.sh`

```bash
#!/usr/bin/env bash
for ((i = 0; i < 10; i++)); do
  printf 'Test\n'
  sleep 1
done
```

`$ nu ./nushell.nu`

```nushell
#!/usr/bin/env -S nu --stdin
def main [] {
  run-external bash ./tmp.sh
  | lines # it will add/trim *newline* automatically
  | each {|line|
    print $"Bash script output: (date now | format date '%Y-%m-%d %H:%M:%S') ($line)"
  }

  return # don't output `empty list`
}
```

Result:
> Bash script output: 2025-01-02 18:41:34 Test
> Bash script output: 2025-01-02 18:41:35 Test
> Bash script output: 2025-01-02 18:41:36 Test
> [...]
